### ISSUE TITLE
Streams, BeginRead/BeginWrite members

### ISSUE THREAD

**Issue by agnauck** [dotnet/coreclr#2952]
it looks like the async functions in Streams are based on BeginWrite/BeginRead internal.
It would great to have the "old" async BeginRead/Write functions available in the API.
There is still lots of code out which relies on them and would be much easier to get ported to the coreclr. Also sometimes it just performs better.

**Comment by stephentoub** [comment]
cc: @terrajobst 

Thanks for the comments, @agnauck.  

Stream in mscorlib has these methods as an implementation detail to remain consistent with other copies of mscorlib and to support other platforms where these are still exposed publically, but they don't exist in other versions of Stream, like the one used for CoreRT (https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/Stream.cs).  And when you look at other Stream-derived types in corefx, many of these Begin/End methods have already been removed, and others are only there temporarily until we have the time to cull them out of the source.

We don't have any plans to bring them back.  They were part of a complicated model of asynchrony that was difficult to implement and difficult to consume. With the async/await language support for Tasks and with the framework firmly moving towards Task-based asynchrony, we don't have any intention of bringing these Begin/End methods back.  

We do realize that in some cases porting of existing code would be easier if these methods were present, but at the same time we're having to walk a fine line between making it easy to port code while also ensuring that the framework is modern and encourages good patterns for the future. If you really find that porting your applications is blocked by lack of Stream.Begin/EndRead/Write, you can create your own simple extension method wrappers on top of ReadAsync/WriteAsync, with helpers like those here: https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Threading/Tasks/TaskToApm.cs.  That should help you to port code expecting that pattern, and then once you're ported you can migrate your code to the newer designs and eventually remove these helpers.

**Comment by agnauck** [comment]
@stephentoub I mostly agree with you.
I forgot to mention that I work with NetworkStream and SslStream. For those streams in high performance environments APM is a very common pattern. Because they are just wrappers around async socket API.

When I use the TaskToApm functions you linked above I get deadlocks. Maybe related to this:
https://github.com/dotnet/corefx/commit/1c5957c4e80a140a92a1b4d11bcc32e106cbc650

**Comment by stephentoub** [comment]
> When I use the TaskToApm functions you linked above I get deadlocks

Can you share a repro?  Please open a separate issue in the corefx repo related to that, and we can help investigate.

> Because they are just wrappers around async socket API.

If performance is critical to you, you don't want to use the socket's APM implementation, either... you want to use the SocketAsyncEventArgs-based APIs that let you reuse the same instances without doing a bunch of allocations on each operation.

**Comment by agnauck** [comment]
> Can you share a repro? Please open a separate issue in the corefx repo related to that, and we can help investigate.

I will run more test of my code to verify where the problem exactly is. Also run a test with the "full" .NET framework 4.6. If there is a problem in corefx I will provide a test case.


> If performance is critical to you, you don't want to use the socket's APM implementation, either... you want to use the SocketAsyncEventArgs-based APIs that let you reuse the same instances without doing a bunch of allocations on each operation.

correct. We have to reconsider the whole design. If it comes to performance its 1) SocketAsyncEventArgs 2) APM 3) TAP.

**Comment by agnauck** [comment]
> Can you share a repro? Please open a separate issue in the corefx repo related to that, and we can help investigate.

#https://github.com/dotnet/corefx/issues/6036

**Comment by stephentoub** [comment]
Thanks.
